services/motion/localizer.go

```diff
@@ -51,31 +55,214 @@ func NewMovementSensorLocalizer(ms movementsensor.MovementSensor, origin *geo.Po
+		return errors.New("Real can't be NaN")
+	}
+
+	return nil
+}
+
+func LogGP(gp geo.Point, before, after time.Time) error {
+	b, err := json.Marshal(GP{GeoPoint: GeoPoint{Lat: gp.Lat(), Lng: gp.Lng()}, Type: "geopoint", StartedAt: before, CompletedAt: after})
+	if err != nil {
+		return err
+	}
+
+	return recordSensorJson(b)
+}
+
+func LogP(p spatialmath.Pose, before, after time.Time) error {
+	b, err := json.Marshal(P{Pose: p, Type: "pose", StartedAt: before, CompletedAt: after})
+	if err != nil {
+		return err
+	}
+	return recordSensorJson(b)
+}
+
+func logProp(p movementsensor.Properties, before, after time.Time) error {
+	b, err := json.Marshal(Prop{Properties: p, Type: "properties", StartedAt: before, CompletedAt: after})
+	if err != nil {
+		return err
+	}
+	return recordSensorJson(b)
+}
+
+func logCH(headingLeft float64, before, after time.Time) error {
+	b, err := json.Marshal(CH{CompassHeading: headingLeft, Type: "compass_heading", StartedAt: before, CompletedAt: after})
+	if err != nil {
+		return err
+	}
+	return recordSensorJson(b)
+}
+
+func logO(o spatialmath.Orientation, before, after time.Time) error {
+	b, err := json.Marshal(O{Orientation: o, Type: "orientation", StartedAt: before, CompletedAt: after})
+	if err != nil {
+		return err
+	}
+	return recordSensorJson(b)
+}
+
+func recordSensorJson(b []byte) error {
+	homedir, err := os.UserHomeDir()
+	if err != nil {
+		return err
+	}
+
+	f, err := os.OpenFile(path.Join(homedir, "localizer.jsonl"), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	if _, err := f.Write(append(b, '\n')); err != nil {
+		return err
+	}
+	return nil
+}
+
 // CurrentPosition returns a movementsensor's current position.
 func (m *movementSensorLocalizer) CurrentPosition(ctx context.Context) (*referenceframe.PoseInFrame, error) {
+	before := time.Now()
 	gp, _, err := m.Position(ctx, nil)
+	after := time.Now()
 	if err != nil {
 		return nil, err
 	}
+
+	if err := LogGP(*gp, before, after); err != nil {
+		return nil, err
+	}
+
+	if err := ValidateGeopoint(gp); err != nil {
+		return nil, err
+	}
 	var o spatialmath.Orientation
+	before = time.Now()
 	properties, err := m.Properties(ctx, nil)
+	after = time.Now()
 	if err != nil {
 		return nil, err
 	}
+
+	if err := logProp(*properties, before, after); err != nil {
+		return nil, err
+	}
+
 	switch {
 	case properties.CompassHeadingSupported:
+		before = time.Now()
 		headingLeft, err := m.CompassHeading(ctx, nil)
+		after = time.Now()
 		if err != nil {
 			return nil, err
 		}
+		if err := logCH(headingLeft, before, after); err != nil {
+			return nil, err
+		}
+
+		if math.IsNaN(headingLeft) {
+			return nil, errors.New("heading can't be NaN")
+		}
 		// CompassHeading is a left-handed value. Convert to be right-handed. Use math.Mod to ensure that 0 reports 0 rather than 360.
+		// NOTE: I still don't understand how to think about handedness when it comes to compass heading.
```

Comment:
Left-handed systems rotate clockwise around the axis of rotation. Right-handed systems rotate counterclockwise.